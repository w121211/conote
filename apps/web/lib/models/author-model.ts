import { Author } from '@prisma/client'
import {
  Author as GQLAuthor,
  AuthorMeta as GQLAuthorMeta,
  AuthorInput as GQLAuthorInput,
} from 'graphql-let/__generated__/__types__'
import prisma from '../prisma'

// type Job = {
//   name: string
//   org: string
//   startedAt?: Date
//   endedAt?: Date
// }

type AuthorMeta = {
  type: 'ORG' | 'PERSON'
  job?: string // eg youtuber, analylist
  org?: string
  sites: [string, string][] // [site_url, site_name]
}

export const AuthorModel = {
  async get(id?: string, name?: string): Promise<GQLAuthor | null> {
    let author: Author | null
    if (id) {
      author = await prisma.author.findUnique({
        where: { id },
      })
    } else if (name) {
      author = await prisma.author.findUnique({
        where: { name },
      })
    } else {
      throw 'Need either id or name to get author'
    }
    if (author) {
      return this.toGQLAuthor(author)
    }
    return null
  },

  async getAll(): Promise<Author[]> {
    console.log('Retreiving all authors from database...')

    let authors: Author[] = []
    let cursor: string | undefined = undefined

    // eslint-disable-next-line no-constant-condition
    while (true) {
      // eslint-disable-next-line no-await-in-loop
      const res: Author[] = await prisma.author.findMany({
        take: 100,
        skip: cursor ? 1 : undefined, // skip cursor itself
        orderBy: { id: 'asc' },
        cursor: cursor ? { id: cursor } : undefined,
      })
      if (res.length === 0) {
        break
      }
      authors = authors.concat(res)
      cursor = res[res.length - 1].id
    }
    return authors
  },

  async create(input: GQLAuthorInput): Promise<Author> {
    const { name, type, job, org, sites } = input
    const meta: AuthorMeta = {
      type,
      job: job ?? undefined,
      org: org ?? undefined,
      sites: sites.map(e => [e.url, e.name]),
    }
    return prisma.author.create({
      data: {
        name,
        meta,
      },
    })
  },

  toGQLAuthor(author: Author): GQLAuthor {
    const { meta } = author
    return {
      ...author,
      meta: this.toGQLAuthorMeta(meta as unknown as AuthorMeta),
    }
  },

  toGQLAuthorMeta(meta: AuthorMeta): GQLAuthorMeta {
    const { sites } = meta
    return {
      ...meta,
      sites: sites.map(([url, name]) => {
        return { url, name }
      }),
    }
  },

  async update(id: string, input: GQLAuthorInput): Promise<Author> {
    const { name, type, job, org, sites } = input
    const meta: AuthorMeta = {
      type,
      job: job ?? undefined,
      org: org ?? undefined,
      sites: sites.map(e => [e.url, e.name]),
    }
    return prisma.author.update({
      data: {
        name,
        meta,
      },
      where: { id },
    })
  },
}
