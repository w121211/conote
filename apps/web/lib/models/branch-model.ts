if (process.env.APP_BRANCH_NAME === undefined) {
  throw new Error('Not found APP_BRANCH_NAME in env')
}

const BRANCH_NAME = process.env.APP_BRANCH_NAME

export function getBranchName(): string {
  return BRANCH_NAME
}
