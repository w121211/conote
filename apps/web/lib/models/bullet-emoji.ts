import { BulletEmoji, BulletEmojiCount, BulletEmojiLike, LikeChoice } from '@prisma/client'
import { EmojiCode } from 'graphql-let/__generated__/__types__'
import prisma from '../prisma'
import { EmojiLike } from './emoji-like'

export const BulletEmojiModel = {
  async create({ bulletId, code, userId }: { bulletId: string; code: EmojiCode; userId: string }): Promise<{
    emoji: BulletEmoji & { count: BulletEmojiCount }
    like: BulletEmojiLike
  }> {
    const found = await prisma.bulletEmoji.findUnique({
      where: { bulletId_code: { bulletId, code } },
    })
    if (found) {
      // emoji existed, upsert like
      const { like, count } = await this.upsertLike({ choice: LikeChoice.UP, bulletEmojiId: found.id, userId })
      return {
        emoji: { ...found, count },
        like,
      }
    }
    // emoji not exist, create emoji & like
    const emoji = await prisma.bulletEmoji.create({
      data: {
        bullet: { connect: { id: bulletId } },
        code,
        count: { create: {} },
      },
    })
    const { like, count } = await this.upsertLike({ choice: LikeChoice.UP, bulletEmojiId: emoji.id, userId })
    return {
      emoji: { ...emoji, count },
      like,
    }
  },

  async upsertLike({
    choice,
    bulletEmojiId,
    userId,
  }: {
    choice: LikeChoice
    bulletEmojiId: string
    userId: string
  }): Promise<{
    like: BulletEmojiLike
    count: BulletEmojiCount
  }> {
    const curCount = await prisma.bulletEmojiCount.findUnique({
      where: { bulletEmojiId },
    })
    if (curCount === null) {
      throw 'Database unexpected error'
    }

    const curLike = await prisma.bulletEmojiLike.findUnique({
      where: {
        userId_bulletEmojiId: { userId, bulletEmojiId },
      },
    })
    const nextLike = curLike
      ? await prisma.bulletEmojiLike.update({
          data: { choice },
          where: { id: curLike.id },
        })
      : await prisma.bulletEmojiLike.create({
          data: {
            choice,
            bulletEmoji: { connect: { id: bulletEmojiId } },
            user: { connect: { id: userId } },
          },
        })

    const { dUp, dDown } = EmojiLike.compare(nextLike, curLike ?? undefined)
    const nextCount = await prisma.bulletEmojiCount.update({
      data: {
        nUps: curCount.nUps + dUp,
        nDowns: curCount.nDowns + dDown,
      },
      where: { id: curCount.id },
    })
    return { like: nextLike, count: nextCount }
  },
}
