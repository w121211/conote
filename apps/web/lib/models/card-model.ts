import { Card, Link, Sym, SymType } from '@prisma/client'
import { FetchClient } from '../fetcher/fetch-client'
import prisma from '../prisma'
import { CardStateModel, CardStateParsed } from './card-state-model'
import { LinkService } from './link-model'
import { SymModel } from './sym-model'

export type CardPrarsed = Omit<Card, 'meta'> & {
  meta: CardMeta
  sym: Sym
  state: CardStateParsed | null
  link: Link | null
}

export type CardMeta = {
  template?: 'webpage' | 'ticker' | 'topic' | 'vs'
  redirects?: string[]
  duplicates?: string[]
  url?: string
  author?: string
  publishedAt?: string
  description?: string
  keywords?: string[]
  // lang?: string
  title?: string
  tickers?: string[]
}

export type RowCard = Card & { sym: Sym; link: Link | null }

export const CardModel = {
  async get(id: string): Promise<null | CardPrarsed> {
    const card = await prisma.card.findUnique({
      where: { id },
      include: {
        link: true,
        sym: true,
      },
    })
    if (card) {
      return {
        ...card,
        meta: card.meta as unknown as CardMeta,
        state: await CardStateModel.getLastCardState(card.id),
      }
    }
    return null
  },

  /**
   * Get card by symbol
   *
   * @throw symbol parse error
   */
  async getBySymbol(symbol: string): Promise<CardPrarsed | null> {
    const parsed = SymModel.parse(symbol)
    const sym = await prisma.sym.findUnique({
      where: { symbol: parsed.symbol },
      include: { card: { include: { link: true } } },
    })
    if (sym?.card) {
      const { card, ...restSym } = sym
      const state = await CardStateModel.getLastCardState(card.id)
      // if (state === null) {
      //   throw '[conote-web] symbol-card state cannot be null'
      // }
      return {
        ...card,
        meta: card.meta as unknown as CardMeta,
        state,
        sym: restSym,
      }
    }
    return null
  },

  /**
   * Get or create a webpage-card by URL, if the URL not found in database, will try to scrape and then store
   * Newly created webpage-card will not have card-state and remains null.
   */
  async getOrCreateByUrl({
    scraper,
    url,
  }: {
    scraper?: FetchClient
    url: string
  }): Promise<CardPrarsed & { link: Link }> {
    const [link, { fetchResult }] = await LinkService.getOrCreateLink({ scraper, url })
    if (link.card) {
      const { card, ...linkRest } = link
      return {
        ...card,
        link: linkRest,
        meta: card.meta as unknown as CardMeta,
        state: await CardStateModel.getLastCardState(card.id),
      }
    }

    // Card not found, create one
    const meta: CardMeta = {
      template: 'webpage',
      url: link.url,
      author: fetchResult?.authorName,
      publishedAt: fetchResult?.date,
      description: fetchResult?.description,
      keywords: fetchResult?.keywords,
      title: fetchResult?.title,
      tickers: fetchResult?.tickers,
    }
    const card = await prisma.card.create({
      data: {
        link: { connect: { id: link.id } },
        meta,
        sym: {
          create: {
            type: SymType.URL,
            symbol: LinkService.toSymbol(link),
          },
        },
      },
      include: { sym: true },
    })
    return {
      ...card,
      link,
      meta,
      state: null,
    }
  },

  parse(card: RowCard, cardState: CardStateParsed | null = null): CardPrarsed {
    const { link, sym } = card
    return {
      ...card,
      meta: card.meta as unknown as CardMeta,
      state: cardState,
      sym,
      link,
    }
  },
}
