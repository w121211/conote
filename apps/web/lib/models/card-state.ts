import { CardState } from '@prisma/client'
import { NodeBody, NodeChange, TreeNode, TreeService } from '@conote/docdiff'
import { Bullet } from '../../components/bullet/bullet'
import prisma from '../prisma'

export type CardStateBody = {
  prevStateId: string | null // initial state
  changes: NodeChange<Bullet>[]
  value: TreeNode<Bullet>[]
  // sourceCardId: string | null
  fromCardId?: string // doc is created from which during editing, not strong binding
}

export type CardStateParsed = Omit<CardState, 'body'> & {
  body: CardStateBody
}

export const CardStateModel = {
  _encodeBody() {
    throw 'Not implemented'
  },

  _decodeBody() {
    throw 'Not implemented'
  },

  async get(id: string): Promise<CardStateParsed | null> {
    const state = await prisma.cardState.findFirst({
      where: { id },
    })
    return state ? this.parse(state) : null
  },

  async getLastCardState(cardId: string): Promise<CardStateParsed | null> {
    const state = await prisma.cardState.findFirst({
      where: { cardId },
      orderBy: { createdAt: 'desc' },
    })
    return state ? this.parse(state) : null
  },

  parse(state: CardState): CardStateParsed {
    return {
      ...state,
      body: state.body as unknown as CardStateBody,
    }
  },
}
