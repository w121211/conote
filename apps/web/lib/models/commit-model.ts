import {
  Commit,
  Discuss,
  Note,
  NoteDoc,
  NoteDraft,
  PrismaPromise,
  Sym,
} from '@prisma/client'
import cuid from 'cuid'
import { isNumber } from 'lodash'
import { NoteDocContent, NoteDocMeta } from '../interfaces'
import prisma from '../prisma'
import { getContentDiff, metaDiff } from './change-model'
import { noteDocModel } from './note-doc-model'
import { noteDraftModel } from './note-draft-model'
import { symModel } from './sym-model'

/**
 *
 */
async function validateDraft(id: string, userId: string) {
  const draft = await prisma.noteDraft.findUnique({
      where: { id },
    }),
    sym =
      draft &&
      (await prisma.sym.findUnique({
        where: { symbol: draft.symbol },
      })),
    curDoc =
      sym &&
      // TODO: noteDocModel.getCurrentNoteDoc(branchId, symId)
      (await prisma.noteDoc.findFirst({
        where: {
          branchId: draft.branchId,
          symId: sym.id,
          status: 'MERGE',
        },
        orderBy: { updatedAt: 'desc' },
      })),
    note =
      draft &&
      sym &&
      (await prisma.note.findUnique({
        where: {
          branchId_symId: {
            branchId: draft.branchId,
            symId: sym.id,
          },
        },
      })),
    draftParsed = draft && noteDraftModel.parse(draft),
    newJoinedDiscussIds =
      draftParsed &&
      draftParsed.content.discussIds.filter(e => e.commitId === undefined),
    newJoinedDiscusses =
      newJoinedDiscussIds &&
      (await prisma.$transaction(
        newJoinedDiscussIds.map(e =>
          prisma.discuss.findUnique({
            where: { id: e.discussId },
          }),
        ),
      ))

  // console.log(`existing discuss ${existingDiscusses?.length}`)

  if (draft === null) {
    throw new Error('Some draftId does not exist.')
  }
  if (draftParsed === null) {
    throw new Error('Some draftId does not exist.')
  }
  if (draft.fromDocId && (curDoc === null || draft.fromDocId !== curDoc.id)) {
    throw new Error('FromDoc is not the latest doc of this note.')
  }
  if (draft.fromDocId === null && curDoc !== null) {
    throw new Error('FromDoc is not the latest doc of this note.')
  }
  if (sym && note === null) {
    throw new Error('Unexpected error: found sym but not found note')
  }
  if (draft.userId !== userId) {
    throw new Error('User is not the owner of the draft')
  }
  // TODO: check if the draft is the same as curDoc
  // if(draft.fromDocId) {
  //   const domainChanged = draft.domain === curDoc!.domain
  //   const metaChanged = metaDiff(curDoc!.meta as unknown as NoteDocMeta, draftParsed.meta)
  //   const contentChanged = getContentDiff(curDoc!.content as unknown as NoteDocContent, draftParsed.content)
  //   if (!domainChanged && !metaChange && !contentChange)
  // }

  return {
    draft,
    draftParsed,
    fromDoc: curDoc,
    sym,
    note,
    newJoinedDiscusses,
  }
}

/**
 *
 */
export async function commitNoteDrafts(
  draftIds: string[],
  userId: string,
): Promise<{
  symbol_symId: Record<string, string>
  commit: Commit
  notes: (Note & {
    discusses: Discuss[]
    sym: Sym
  })[]
  noteDocs: NoteDoc[]
}> {
  const drafts: NoteDraft[] = [],
    symbol_symId: Record<string, string> = {},
    symbol_noteId: Record<string, string> = {},
    promises: PrismaPromise<any>[] = [],
    commitId = cuid()

  // Loop first run
  for (const id of draftIds) {
    const { draft, draftParsed, fromDoc, sym, note, newJoinedDiscusses } =
        await validateDraft(id, userId),
      { branchId, linkId, symbol } = draft,
      symId = sym ? sym.id : cuid(),
      noteId = note ? note.id : cuid()

    drafts.push(draft)
    symbol_symId[symbol] = symId
    symbol_noteId[symbol] = noteId

    // Create Sym and Note if not found
    if (sym === null) {
      const { type } = symModel.parse(symbol)
      promises.push(
        prisma.sym.create({
          data: {
            id: symId,
            type,
            symbol,
            notes: {
              create: {
                id: noteId,
                branch: { connect: { id: branchId } },
                link: linkId ? { connect: { id: linkId } } : undefined,
              },
            },
          },
        }),
      )
    }

    // Connect each new-joined-discuss && whose status is "DRAFT" to Note
    // and update its status to "ACITVE"
    if (newJoinedDiscusses) {
      newJoinedDiscusses.forEach(e => {
        if (e?.status === 'DRAFT') {
          promises.push(
            prisma.discuss.update({
              data: {
                notes: { connect: { branchId_symId: { branchId, symId } } },
                status: 'ACTIVE',
              },
              where: { id: e.id },
            }),
          )
        } else {
          promises.push(
            prisma.discuss.update({
              data: {
                notes: { connect: { branchId_symId: { branchId, symId } } },
              },
              where: { id: e?.id },
            }),
          )
        }
      })
    }
  }

  // Loop second run: for each draft, create doc and connect to the note
  for (const e of drafts) {
    const draftParsed = noteDraftModel.parse(e),
      symId = symbol_symId[e.symbol]

    if (symId === null) throw new Error('Unexpected error: symId === null')

    promises.push(
      // Also update symbol_symId in this function
      noteDocModel.createDoc(
        draftParsed,
        commitId,
        symId,
        userId,
        symbol_symId,
      ),
    )

    // For each draft, connect draft to commit and update its status to commit
    promises.push(
      prisma.noteDraft.update({
        where: { id: e.id },
        data: { commit: { connect: { id: commitId } }, status: 'COMMIT' },
      }),
    )
  }

  // Create Commit and get or create Sym, Note and NoteDoc
  const [commit] = await prisma.$transaction([
    prisma.commit.create({
      data: { id: commitId, user: { connect: { id: userId } } },
    }),
    ...promises,
  ])

  const noteDocs = await prisma.noteDoc.findMany({
    where: { commitId: commit.id },
  })

  const notes = (
    await prisma.$transaction(
      drafts.map(e =>
        prisma.note.findUnique({
          where: {
            branchId_symId: {
              branchId: e.branchId,
              symId: symbol_symId[e.symbol],
            },
          },
          include: {
            discusses: true,
            sym: true,
          },
        }),
      ),
    )
  ).filter(
    (
      e,
    ): e is Note & {
      discusses: Discuss[]
      sym: Sym
    } => e !== null,
  )

  return { symbol_symId, commit, notes, noteDocs }
}
