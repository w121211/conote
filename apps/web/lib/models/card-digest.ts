import { Sym } from '@prisma/client'
import { CardDigest as GQLCardDigest } from 'graphql-let/__generated__/__types__'
import { NodeBody, NodeChange, TreeNode, TreeService } from '@conote/docdiff'
import { Bullet } from '../../components/bullet/bullet'
import prisma from '../prisma'
import { CardMeta, CardModel } from './card'
import { CardStateModel } from './card-state'
import { RowCardState, RowCommit } from './commit'

type CardDigest = {
  cardId: string
  cardMeta: CardMeta
  commitId: string
  fromCardId?: string
  picks: string[]
  sym: Sym
  updatedAt: Date
}

export const CardDigestModel = {
  // fromCard(card: CardPrarsed, subs: GQLCardDigest[]): GQLCardDigest {
  //   if (subs.length === 0) {
  //     throw 'Require 1 or more sub-digests to create a null-state-card digest'
  //   }
  //   const { id: cardId, sym, meta: cardMeta } = card
  //   return {
  //     commitId: subs[0].commitId, // TODO: temp fill
  //     cardId,
  //     cardMeta,
  //     sym,
  //     title: sym.symbol,
  //     picks: [],
  //     subs,
  //     updatedAt: subs[0].updatedAt, // TODO:  temp fill
  //   }
  // },

  fromCardState(state: RowCardState): Omit<CardDigest, 'children'> {
    const { meta: cardMeta, sym } = CardModel.parse(state.card)
    const { body, cardId, commitId, createdAt, updatedAt } = CardStateModel.parse(state)
    return {
      commitId,
      cardId,
      cardMeta,
      fromCardId: body.fromCardId,
      sym,
      picks: body.changes
        .filter((e): e is NodeChange<Bullet> & { data: Bullet } => e.data !== undefined && e.type === 'insert')
        .map(e => e.data.head),
      updatedAt,
    }
  },

  fromCommit(commit: RowCommit): TreeNode<CardDigest>[] {
    const nodes = TreeService.fromList(commit.cardStates.map(e => this.toNodeBody(this.fromCardState(e))))
    return nodes
  },

  async _getLatest(afterCommitId?: string): Promise<TreeNode<CardDigest>[]> {
    // const maxDate = dayjs().startOf('d').subtract(7, 'd')
    const commits = await prisma.commit.findMany({
      // where: { createdAt: { gte: maxDate.toDate() } },
      orderBy: { updatedAt: 'desc' },
      cursor: afterCommitId ? { id: afterCommitId } : undefined,
      take: 20,
      skip: afterCommitId ? 1 : 0,
      include: { cardStates: { include: { card: { include: { sym: true, link: true } } } } },
    })
    const digests = commits
      .map(e => this.fromCommit(e))
      .reduce((acc, cur) => {
        const filtered = cur.filter(e => acc.find(f => f.data?.sym.id === e.data?.sym.id) === undefined) // only keep the latest digest for the same symbol
        return [...acc, ...filtered] // flatten
      }, [])
    return digests
  },

  async getLatest(afterCommitId?: string): Promise<GQLCardDigest[]> {
    const nodes = await this._getLatest(afterCommitId)
    return nodes.map(e => this.toGQLCardDigest(e))
  },

  toGQLCardDigest(node: TreeNode<CardDigest>): GQLCardDigest {
    if (node.data === undefined) {
      throw 'node.data === undefined'
    }
    const children = TreeService.toList(node.children)
    return {
      ...node.data,
      subSyms: children.map(e => e.data?.sym).filter<Sym>((e): e is Sym => e !== undefined),
    }
  },

  toNodeBody(digest: CardDigest): NodeBody<CardDigest> {
    const { cardId, fromCardId, updatedAt } = digest
    return {
      cid: cardId, // since parent-cid is bind with card-id
      data: digest,
      parentCid: fromCardId ?? TreeService.tempRootCid,
      index: updatedAt.getTime(),
    }
  },
}
