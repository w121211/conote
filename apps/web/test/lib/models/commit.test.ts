/* eslint-disable no-await-in-loop */
import prisma from '../../../lib/prisma'
import { testHelper } from '../../test-helpers'
import { commitNoteDrafts } from '../../../lib/models/commit-model'
import { mockNoteDrafts } from '../../__mocks__/mock-note-draft'
import { mockUsers } from '../../__mocks__/mock-user'
import { mockNoteDocs } from '../../__mocks__/mock-note-doc'

beforeAll(async () => {
  // Writing required data into database
  await prisma.$queryRaw`TRUNCATE "Author", "Branch", "User" CASCADE;`
  await testHelper.createBranches(prisma)
  await testHelper.createUsers(prisma)

  // console.log('Setting up a fetch-client')
  // fetcher = new FetchClient(resolve(__dirname, '.cache.fetcher.json'))
})

beforeEach(async () => {
  await prisma.$queryRaw`TRUNCATE "Commit", "Discuss", "Note", "NoteDoc", "NoteDraft", "Link", "Sym" CASCADE;`
  await testHelper.createDiscusses(prisma)
})

afterEach(async () => {
  // Bug: comment out to avoid rerun loop  @see https://github.com/facebook/jest/issues/2516
  // fetcher.dump()
  await prisma.$disconnect()
})

describe('commitNoteDrafts()', () => {
  /**
   * Validations
   *
   * Cases which throws error:
   * - [x] user is not the owner of draft
   * - [x] draftId cannot be found
   * - [x] fromDoc is empty but there is a latest doc
   * - [x] fromDoc does not match
   *
   */

  it('throws if user is not the owner of draft', async () => {
    await testHelper.createNoteDrafts(prisma, [
      mockNoteDrafts[0],
      mockNoteDrafts[1],
    ])
    await expect(async () => {
      await commitNoteDrafts(
        [mockNoteDrafts[0].id, mockNoteDrafts[1].id],
        mockUsers[0].id,
      )
    }).rejects.toThrowErrorMatchingInlineSnapshot(
      `"User is not the owner of the draft"`,
    )
  })

  it('throws if draftId cannot be found', async () => {
    await testHelper.createNoteDrafts(prisma, [mockNoteDrafts[0]])
    await expect(async () => {
      await commitNoteDrafts(
        [mockNoteDrafts[0].id, 'a-not-found-draft-id'],
        mockUsers[0].id,
      )
    }).rejects.toThrowErrorMatchingInlineSnapshot(
      `"Some draftId does not exist."`,
    )
  })

  it('throws if fromDoc is empty but there is a latest doc ', async () => {
    await testHelper.createNoteDrafts(prisma, [mockNoteDrafts[0]])
    await testHelper.createCommit(prisma)
    await expect(async () => {
      await commitNoteDrafts([mockNoteDrafts[0].id], mockUsers[0].id)
    }).rejects.toThrowErrorMatchingInlineSnapshot(
      `"FromDoc is not the latest doc of this note."`,
    )
  })

  /**
   * Executions
   *
   * Cases:
   * - [x] (1) from-doc is null, create sym, note
   * - [x] (2) from-doc is not null
   * - [x] mix of (1) and (2) (commit drafts with null and not null for from-doc)
   * - [x] got discuss ids, should only set this-draft-created-discusses to "ACTIVE"
   * - [ ] commit a web-note (ie, got link-id) if from-doc is null => not equals to case (1),
   *       need one extra step: connect sym & link
   * - [x] commit a web-note if from-doc is not null => equals to case (2)
   *
   */

  it('create sym, note if from-doc not exist', async () => {
    await testHelper.createNoteDrafts(prisma, [
      mockNoteDrafts[0],
      {
        ...mockNoteDrafts[1],
        userId: mockUsers[0].id,
      },
    ])
    const { commit, notes, noteDocs, symbol_symId } = await commitNoteDrafts(
      [mockNoteDrafts[0].id, mockNoteDrafts[1].id],
      mockUsers[0].id,
    )
    expect(Object.keys(symbol_symId)).toMatchInlineSnapshot(`
      Array [
        "[[Apple]]",
        "[[Google]]",
      ]
    `)
    expect(notes.map(e => e.sym.symbol)).toMatchInlineSnapshot(`
      Array [
        "[[Apple]]",
        "[[Google]]",
      ]
    `)
    expect(noteDocs.map(e => e.status)).toMatchInlineSnapshot(`
      Array [
        "CANDIDATE",
        "CANDIDATE",
      ]
    `)
    const updatedDrafts = await prisma.noteDraft.findMany({
      where: { commit: { id: commit.id } },
    })
    expect(updatedDrafts.map(e => [e.symbol, e.status])).toMatchInlineSnapshot(`
      Array [
        Array [
          "[[Google]]",
          "COMMIT",
        ],
        Array [
          "[[Apple]]",
          "COMMIT",
        ],
      ]
    `)
    const noteDocHistory = await prisma.noteDoc.findMany({
      where: { note: { id: notes[0].id } },
      take: 3,
    })
    expect(noteDocHistory.map(e => e.domain)).toMatchInlineSnapshot(`
      Array [
        "domain0",
      ]
    `)
  })

  it('create noteDoc if from-doc exists', async () => {
    // Start with a draft without from-doc and simulate the first-commit to get the note-doc as from-doc
    // This way removes the hassle of caring the commit process
    await testHelper.createNoteDrafts(prisma, [mockNoteDrafts[0]])
    const { sym, noteDoc } = await testHelper.createCommit(prisma)
    await testHelper.createNoteDrafts(prisma, [
      {
        ...mockNoteDrafts[0],
        id: 'mock-draft-0-1_from-empty',
        symbol: sym.symbol,
        userId: mockUsers[1].id,
        fromDocId: noteDoc.id,
        domain: noteDoc.domain,
      },
    ])

    expect(
      await prisma.noteDraft.findUnique({
        where: { id: 'mock-draft-0-1_from-empty' },
        select: { fromDocId: true },
      }),
    ).toMatchInlineSnapshot(`
      Object {
        "fromDocId": "mock-doc-1_merge",
      }
    `)

    const { commit, notes, noteDocs, symbol_symId } = await commitNoteDrafts(
      ['mock-draft-0-1_from-empty'],
      mockUsers[1].id,
    )
    expect(Object.keys(symbol_symId)).toMatchInlineSnapshot(`
      Array [
        "[[Apple]]",
      ]
    `)
    expect(notes.map(e => e.sym.symbol)).toMatchInlineSnapshot(`
      Array [
        "[[Apple]]",
      ]
    `)
    expect(noteDocs.map(e => e.status)).toMatchInlineSnapshot(`
      Array [
        "CANDIDATE",
      ]
    `)

    const updatedDrafts = await prisma.noteDraft.findMany({
      where: { commit: { id: commit.id } },
    })
    expect(updatedDrafts.map(e => [e.symbol, e.status])).toMatchInlineSnapshot(`
      Array [
        Array [
          "[[Apple]]",
          "COMMIT",
        ],
      ]
    `)

    const noteDocHistory = await prisma.noteDoc.findMany({
      where: { note: { id: notes[0].id } },
      take: 3,
      orderBy: { createdAt: 'asc' },
    })
    expect(
      noteDocHistory.map(e => ({
        domain: e.domain,
        status: e.status,
        fromDocId: e.fromDocId,
      })),
    ).toMatchInlineSnapshot(`
      Array [
        Object {
          "domain": "domain0",
          "fromDocId": null,
          "status": "MERGE",
        },
        Object {
          "domain": "domain0",
          "fromDocId": "mock-doc-1_merge",
          "status": "CANDIDATE",
        },
      ]
    `)
  })

  it('got discuss ids, should only set this-draft-created-discusses to "ACTIVE"', async () => {
    await testHelper.createNoteDrafts(prisma, [mockNoteDrafts[1]])

    expect(
      (await prisma.discuss.findMany()).map(e => ({
        id: e.id,
        userId: e.userId,
        status: e.status,
      })),
    ).toMatchInlineSnapshot(`
      Array [
        Object {
          "id": "mock-discuss-0_active",
          "status": "ACTIVE",
          "userId": "testuser0",
        },
        Object {
          "id": "mock-discuss-1_draft",
          "status": "DRAFT",
          "userId": "testuser1",
        },
        Object {
          "id": "mock-discuss-2_archive",
          "status": "ARCHIVE",
          "userId": "testuser2",
        },
      ]
    `)
    const result = await commitNoteDrafts(
      [mockNoteDrafts[1].id],
      mockUsers[1].id,
    )
    expect(
      result.notes.map(e => {
        return e.discusses.map(d => ({
          id: d.id,
          userId: d.userId,
          status: d.status,
        }))
      }),
    ).toMatchInlineSnapshot(`
      Array [
        Array [
          Object {
            "id": "mock-discuss-0_active",
            "status": "ACTIVE",
            "userId": "testuser0",
          },
          Object {
            "id": "mock-discuss-1_draft",
            "status": "ACTIVE",
            "userId": "testuser1",
          },
          Object {
            "id": "mock-discuss-2_archive",
            "status": "ARCHIVE",
            "userId": "testuser2",
          },
        ],
      ]
    `)
  })

  it('create multiple noteDocs if one from-doc exists and the other does not', async () => {
    await testHelper.createNoteDrafts(prisma, [mockNoteDrafts[0]])
    await testHelper.createCommit(prisma)
    await testHelper.createNoteDrafts(prisma, [
      {
        ...mockNoteDrafts[0],
        id: 'mock-draft-0-1_from-empty',
        userId: mockUsers[1].id,
        fromDocId: mockNoteDocs[1].id,
      },
      mockNoteDrafts[1],
    ])
    const { commit, notes, noteDocs, symbol_symId } = await commitNoteDrafts(
      ['mock-draft-0-1_from-empty', mockNoteDrafts[1].id],
      mockUsers[1].id,
    )
    expect(Object.keys(symbol_symId)).toMatchInlineSnapshot(`
      Array [
        "[[Apple]]",
        "[[Google]]",
      ]
    `)
    expect(notes.map(e => e.sym.symbol)).toMatchInlineSnapshot(`
      Array [
        "[[Apple]]",
        "[[Google]]",
      ]
    `)
    expect(noteDocs.map(e => ({ fromDocId: e.fromDocId, status: e.status })))
      .toMatchInlineSnapshot(`
      Array [
        Object {
          "fromDocId": "mock-doc-1_merge",
          "status": "CANDIDATE",
        },
        Object {
          "fromDocId": null,
          "status": "CANDIDATE",
        },
      ]
    `)
    const updatedDrafts = await prisma.noteDraft.findMany({
      where: { commit: { id: commit.id } },
    })
    expect(updatedDrafts.map(e => [e.symbol, e.status])).toMatchInlineSnapshot(`
      Array [
        Array [
          "[[Google]]",
          "COMMIT",
        ],
        Array [
          "[[Apple]]",
          "COMMIT",
        ],
      ]
    `)
  })
})
