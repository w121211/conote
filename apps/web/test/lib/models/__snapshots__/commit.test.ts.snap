// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create commit: card-state has card-input 1`] = `
Object {
  "cardStates": Array [
    Object {
      "body": Object {
        "changes": Array [],
        "prevStateId": null,
        "sourceCardId": null,
        "value": Array [
          Object {
            "children": Array [],
            "cid": "0",
            "data": Object {
              "head": "000",
            },
            "index": 0,
            "parentCid": "TEMP_ROOT_CID",
          },
        ],
      },
      "card": Object {
        "linkId": null,
        "meta": Object {},
        "sym": Object {
          "symbol": "$AA",
          "type": "TICKER",
        },
      },
      "cardId": "1",
      "cid": "$AA",
      "commitId": "4",
      "prevId": null,
      "userId": "testuser0",
    },
  ],
  "userId": "testuser0",
}
`;

exports[`create commit: card-state has card-input 2`] = `
Object {
  "cardStates": Array [
    Object {
      "body": Object {
        "changes": Array [],
        "prevStateId": "2",
        "sourceCardId": null,
        "value": Array [
          Object {
            "children": Array [],
            "cid": "0",
            "data": Object {
              "head": "000",
            },
            "index": 0,
            "parentCid": "TEMP_ROOT_CID",
          },
        ],
      },
      "card": Object {
        "linkId": null,
        "meta": Object {},
        "sym": Object {
          "symbol": "$AA",
          "type": "TICKER",
        },
      },
      "cardId": "1",
      "cid": "$AA",
      "commitId": "7",
      "prevId": "2",
      "userId": "testuser0",
    },
  ],
  "userId": "testuser0",
}
`;

exports[`create commit: multi card-state 1`] = `
Object {
  "cardStates": Array [
    Object {
      "body": Object {
        "changes": Array [],
        "prevStateId": null,
        "sourceCardId": "1",
        "value": Array [
          Object {
            "children": Array [],
            "cid": "0",
            "data": Object {
              "head": "000",
            },
            "index": 0,
            "parentCid": "TEMP_ROOT_CID",
          },
        ],
      },
      "card": Object {
        "linkId": null,
        "meta": Object {},
        "sym": Object {
          "symbol": "$BB",
          "type": "TICKER",
        },
      },
      "cardId": "8",
      "cid": "$BB",
      "commitId": "12",
      "prevId": null,
      "userId": "testuser0",
    },
    Object {
      "body": Object {
        "changes": Array [],
        "prevStateId": null,
        "sourceCardId": "1",
        "value": Array [
          Object {
            "children": Array [],
            "cid": "0",
            "data": Object {
              "head": "000",
            },
            "index": 0,
            "parentCid": "TEMP_ROOT_CID",
          },
        ],
      },
      "card": Object {
        "linkId": null,
        "meta": Object {},
        "sym": Object {
          "symbol": "$CC",
          "type": "TICKER",
        },
      },
      "cardId": "9",
      "cid": "$CC",
      "commitId": "12",
      "prevId": null,
      "userId": "testuser0",
    },
  ],
  "userId": "testuser0",
}
`;
